### PR TITLE
[core] update oom message to include task id and retriable info

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

Add task id and whether the task is retriable in the oom error message.

Fix OOM worker exit details truncated due to the link breaks.

Example message, non-retriable actor
==

>>>
(raylet) Memory on the node (IP: 172.17.0.2, ID: a6a17f664d1a71d5ef1f64f49e955c96b0655bd1c37671145ba92181) where the task (actor ID: a019b640a0505d63a35d2c4201000000) was running was 21.81GB / 33.28GB (0.655227), which exceeds the memory usage threshold of 0.65. Ray killed this worker (ID: 3a1aaaecda58df909728e6a7099238577313f0e5b6131eece854bb26) to free up memory. The task/actor is not retriable, which can fail the application. To see more information about memory usage on this node, use `ray logs raylet.out -ip 172.17.0.2`. To see the logs of the worker, use `ray logs worker-3a1aaaecda58df909728e6a7099238577313f0e5b6131eece854bb26*out -ip 172.17.0.2`. 
(raylet) Consider provisioning more memory on this node or reducing task parallelism by requesting more CPUs per task. Set max_restarts and max_task_retries to increase restart counts.To adjust the kill threshold, set the environment variable `RAY_memory_usage_threshold_fraction` when starting Ray. To disable worker killing, set the environment variable `RAY_memory_monitor_interval_ms` to zero.
>>>

Example message, retriable actor
==

>>>
(raylet) Memory on the node (IP: 172.17.0.2, ID: a553fd80555c1d7c8b60c181c775657b4ac5adc96401d3046da06bda) where the task (actor ID: 8aa0fdad03c17cd61ed0042a01000000) was running was 21.73GB / 33.28GB (0.653), which exceeds the memory usage threshold of 0.65. Ray killed this worker (ID: a5ca5c37930acbd3aa51f29120254a48afac5087f0a8a0123652237c) because it was the most recently scheduled task that is retriable; to see more information about memory usage on this node, use `ray logs raylet.out -ip 172.17.0.2`. To see the logs of the worker, use `ray logs worker-a5ca5c37930acbd3aa51f29120254a48afac5087f0a8a0123652237c*out -ip 172.17.0.2`.
(raylet) 
(raylet) Consider provisioning more memory on this node or reducing task parallelism by requesting more CPUs per task. To adjust the kill threshold, set the environment variable `RAY_memory_usage_threshold_fraction` when starting Ray. To disable worker killing, set the environment variable `RAY_memory_monitor_interval_ms` to zero.
>>>

Also fixes the issue where the error message in the frontend is truncated due to link break

Example worker exit detail
==

>>>
E                       ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
E                               class_name: Leaker
E                               actor_id: d924bc18ce770d055e43dc9c01000000
E                               pid: 43308
E                               namespace: f15642fa-b6c2-4511-9a0c-ab22a76903c5
E                               ip: 172.17.0.2
E                       The actor is dead because its worker process has died. Worker exit type: USER_ERROR Worker exit detail: Task was killed due to the node running low on memory. Memory on the node (IP: 172.17.0.2, ID: b883dd430e80ccef1eecd2b3999058aa5a83ed005c3af48d9768da8b) where the task (actor ID: d924bc18ce770d055e43dc9c01000000) was running was 21.92GB / 33.28GB (0.658673), which exceeds the memory usage threshold of 0.65. Ray killed this worker (ID: 9ba2a8ceee91bdda3ecfcde2b3c41f3a37f56b74ab2ab595c9330a02) because it was the most recently scheduled task that is retriable; to see more information about memory usage on this node, use `ray logs raylet.out -ip 172.17.0.2`. To see the logs of the worker, use `ray logs worker-9ba2a8ceee91bdda3ecfcde2b3c41f3a37f56b74ab2ab595c9330a02*out -ip 172.17.0.2`. Consider provisioning more memory on this node or reducing task parallelism by requesting more CPUs per task. To adjust the kill threshold, set the environment variable `RAY_memory_usage_threshold_fraction` when starting Ray. To disable worker killing, set the environment variable `RAY_memory_monitor_interval_ms` to zero.
>>>

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -2945,39 +2945,62 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
               static_cast<float>(system_memory.total_bytes) / 1024 / 1024 / 1024, 2);
           std::stringstream id_ss;
           if (worker_to_kill->GetActorId().IsNil()) {
-            id_ss << "task ID " << worker_to_kill->GetAssignedTaskId();
+            id_ss << "task ID: " << worker_to_kill->GetAssignedTaskId();
           } else {
-            id_ss << "actor ID " << worker_to_kill->GetActorId();
+            id_ss << "actor ID: " << worker_to_kill->GetActorId();
+          }
+
```
Comment (scv119) [inline src/ray/raylet/node_manager.cc:2952]: nit: create helper function for logging stuff?

**Review by scv119 — APPROVED** [review]

**Comment by rkooo567** [comment]
Hmm this error message sounds a bit weird to me

```
Ray killed this worker (ID: a1836fde08972b18c11609d4c6df0496e3ecaae2cef06cc1ef6500fa) because it was the most recently scheduled task that is not retriable - there are no more retriable tasks to kill;
```

1. isn't it exposing too much implementation detail?
2. killed the worker because task is not retriable sounds a bit weird.

Instead, why don't we focus on the consequence? E.g., Ray killed this worker (ID: a1836fde08972b18c11609d4c6df0496e3ecaae2cef06cc1ef6500fa). See <link> for more details about the worker eviction policy under memory pressure. The task/actor <ID> is not retriable, which can fail the application. Set <restart args> to increase restart counts.

**Review by rkooo567 — APPROVED** [review]

**Comment by scv119** [comment]
there are some merge conflicts.

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -3002,6 +2979,64 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
...
+  }
+  std::stringstream oom_kill_details_ss;
+  oom_kill_details_ss
+      << "Memory on the node (IP: " << worker->IpAddress() << ", ID: " << node_id
+      << ") where the task (" << worker->GetIdAsDebugString() << ") was running was "
+      << used_bytes_gb << "GB / " << total_bytes_gb << "GB (" << usage_fraction
+      << "), which exceeds the memory usage threshold of " << usage_threshold
+      << ". Ray killed this worker (ID: " << worker->WorkerId() << ") to free up memory."
+      << is_retriable_ss.str()
+      << " See documentation for more details about the worker killing policy under "
```
[the remark this answers is not in the document]
Comment (clarng) [reply src/ray/raylet/node_manager.cc:3007]: yup once there it a link we will add it, right now there is a todo

**Comment by rkooo567** [comment]
Not sure if it is newly created again, but there's still a merge conflict!

**Comment by clarng** [comment]
Yea a bunch of other changes went in. Probably will try to land the docs https://github.com/ray-project/ray/pull/29341 
and then resolve the all conflicts in one go

**Comment by clarng** [comment]
Superceded by https://github.com/ray-project/ray/pull/29738/files

**Pull request closed without merging**